### PR TITLE
WIP identify suspected cheaters

### PR DESCRIPTION
**Pull request by george-misan**

### CODE REVIEW

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE suspected_cheaters (
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  student_id UUID NOT NULL REFERENCES course_module_completion,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMP WITH TIME ZONE,
+  total_duration VARCHAR(32) NOT NULL,
+  student_average_duration VARCHAR(32) NOT NULL,
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:7]: Create a new table that stores the average for a course

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  total_points VARCHAR(32) NOT NULL student_average_points VARCHAR(32) NOT NULL
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:8]: Missing comma

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE suspected_cheaters (
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  student_id UUID NOT NULL REFERENCES course_module_completion,
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:3]: user_id, REFERENCES users

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+);
+COMMENT ON TABLE suspected_cheaters IS 'This table stores data regarding student that are suspected of cheating in a course.';
+COMMENT ON COLUMN suspected_cheaters.id IS 'A unique, stable identifier for the record.';
+COMMENT ON COLUMN suspected_cheaters.student_id IS 'The id of the student being suspected.';
+COMMENT ON COLUMN suspected_cheaters.created_at IS 'Timestamp when the record was created.';
+COMMENT ON COLUMN suspected_cheaters.deleted_at IS 'Timestamp when the record was deleted. If null, the record is not deleted.';
+COMMENT ON COLUMN suspected_cheaters.total_duration IS 'The total duration the student spend completing the course.';
+COMMENT ON COLUMN suspected_cheaters.student_average_duration IS 'The average total duration other student spent completing the course.';
+COMMENT ON COLUMN suspected_cheaters.total_points IS 'The total points the student acquired in the course.';
+COMMENT ON COLUMN suspected_cheaters.student_average_points IS 'The average total points other students acquired in the course.';
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:18]: missing the trigger set timstamp

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE suspected_cheaters (
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  student_id UUID NOT NULL REFERENCES course_module_completion,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:4]: updated_at column

File: services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
+CREATE TABLE exercise_list_of_suspected_cheaters_exercise_list (
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql:1]: suspected_cheaters_by_exercises

File: services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  student_id UUID NOT NULL REFERENCES course_module_completion,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMP WITH TIME ZONE,
+  exercise_id UUID NOT NULL,
+  duration VARCHAR(32) NOT NULL,
+  student_average_duration VARCHAR(32) NOT NULL,
+  points VARCHAR(32) NOT NULL,
+  student_average_points VARCHAR(32) NOT NULL,
+  attempts VARCHAR(32) NOT NULL,
+  status TEXT NOT NULL
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql:12]: https://rage.github.io/secret-project-331/headless_lms_models/exercises/enum.ActivityProgress.html

File: services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  student_id UUID NOT NULL REFERENCES course_module_completion,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMP WITH TIME ZONE,
+  exercise_id UUID NOT NULL,
+  duration VARCHAR(32) NOT NULL,
+  student_average_duration VARCHAR(32) NOT NULL,
+  points VARCHAR(32) NOT NULL,
+  student_average_points VARCHAR(32) NOT NULL,
+  attempts VARCHAR(32) NOT NULL,
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql:11]: number

File: services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql [hunk 9]
```diff
@@ -0,0 +1,25 @@
+CREATE TABLE exercise_list_of_suspected_cheaters_exercise_list (
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  student_id UUID NOT NULL REFERENCES course_module_completion,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMP WITH TIME ZONE,
+  exercise_id UUID NOT NULL,
+  duration VARCHAR(32) NOT NULL,
+  student_average_duration VARCHAR(32) NOT NULL,
+  points VARCHAR(32) NOT NULL,
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql:9]: number

File: services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql [hunk 10]
```diff
@@ -0,0 +1,25 @@
+CREATE TABLE exercise_list_of_suspected_cheaters_exercise_list (
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  student_id UUID NOT NULL REFERENCES course_module_completion,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMP WITH TIME ZONE,
+  exercise_id UUID NOT NULL,
+  duration VARCHAR(32) NOT NULL,
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql:7]: number

File: services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql [hunk 11]
```diff
@@ -0,0 +1,25 @@
+CREATE TABLE exercise_list_of_suspected_cheaters_exercise_list (
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  student_id UUID NOT NULL REFERENCES course_module_completion,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMP WITH TIME ZONE,
+  exercise_id UUID NOT NULL,
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql:6]: references exercises

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql (hunk 3 above)
Comment (george-misan) [reply services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:3]: please I cant recall, why do we need user_id when we already have student_i?

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql [hunk 12]
```diff
@@ -1,18 +1,46 @@
...
+  average_duration INTEGER NOT NULL,
+  average_points INTEGER NOT NULL
+);
+
+CREATE TRIGGER set_timestamp
+BEFORE UPDATE ON course_student_average
+FOR EACH ROW
+EXECUTE PROCEDURE trigger_set_timestamp();
+
+COMMENT ON TABLE course_student_average IS 'This table stores data regarding the average in a specific course.';
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:40]: Be more specific, tell a bit about the feature this is used for

File: services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql [hunk 13]
```diff
@@ -1,25 +1,62 @@
-CREATE TABLE exercise_list_of_suspected_cheaters_exercise_list (
+CREATE TABLE suspected_cheaters_exercise_list (
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312121500_add_suspected_cheaters_exercise_list.up.sql:1]: Merge the migrations to one. You can add empty space like this:
-- 
--

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/headless-lms/models/src/suspected_cheaters.rs [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[cfg_attr(feature = "ts_rs", derive(TS))]
+pub struct SuspectedCheaters {
+    pub id: Uuid,
+    pub user_id: Uuid,
+    pub created_at: DateTime<Utc>,
+    pub deleted_at: Option<DateTime<Utc>>,
+    pub updated_at: Option<DateTime<Utc>>,
+    pub total_duration: u64, // Represented in milliseconds
```
Comment (nygrenh) [inline services/headless-lms/models/src/suspected_cheaters.rs:11]: total_duration -> total_duration_ms

File: services/headless-lms/models/src/suspected_cheaters.rs [hunk 15]
```diff
@@ -0,0 +1,110 @@
+use crate::prelude::*;
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[cfg_attr(feature = "ts_rs", derive(TS))]
+pub struct SuspectedCheaters {
```
Comment (nygrenh) [inline services/headless-lms/models/src/suspected_cheaters.rs:5]: Name should reflect that this is a result of the analysis

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/headless-lms/models/src/course_module_completions.rs [hunk 16]
```diff
@@ -349,6 +361,31 @@ WHERE course_id = $1
...
+        r"
+    SELECT course_instance_id,
+        SUM(grade) AS total_score,
+        AVG(grade) AS average_score,
+        COUNT(DISTINCT user_id) AS student_count
+    FROM 
+        course_module_completions
+    WHERE 
+        course_id = $1
+        AND deleted_at IS NULL;
```
Comment (nygrenh) [inline services/headless-lms/models/src/course_module_completions.rs:380]: group by

**Review by nygrenh — CHANGES_REQUESTED** [review]

File: services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+  user_id UUID NOT NULL REFERENCES users,
+  created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  updated_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
+  deleted_at TIMESTAMP WITH TIME ZONE,
+  total_duration_seconds INTEGER,
+  total_points INTEGER NOT NULL
+);
+CREATE TRIGGER set_timestamp BEFORE
+UPDATE ON suspected_cheaters FOR EACH ROW EXECUTE PROCEDURE trigger_set_timestamp();
+-- The suspected_cheaters table is a that contains a student that are suspected of cheating because they meet the cheating requirement (i.e. score > threshold && duration > average_duration)
```
Comment (nygrenh) [inline services/headless-lms/migrations/20240312115600_add_suspected_cheaters.up.sql:12]: include in the comment on table

**Review by nygrenh — CHANGES_REQUESTED** [review]

**Review by nygrenh — APPROVED** [review]

**Pull request merged**